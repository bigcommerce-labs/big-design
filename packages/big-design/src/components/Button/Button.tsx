import React, { ComponentPropsWithoutRef, forwardRef, memo, Ref } from 'react';

import { MarginProps } from '../../mixins';
import { ProgressCircle } from '../ProgressCircle';

import { ContentWrapper, LoadingSpinnerWrapper, StyledButton } from './styled';

interface PrivateProps {
  forwardedRef: Ref<HTMLButtonElement>;
}

export interface ButtonProps extends ComponentPropsWithoutRef<'button'>, MarginProps {
  actionType?: 'normal' | 'destructive';
  children?: React.ReactNode;
  iconLeft?: React.ReactNode;
  iconOnly?: React.ReactNode;
  iconRight?: React.ReactNode;
  isLoading?: boolean;
  mobileWidth?: 'auto' | '100%';
  variant?: 'primary' | 'secondary' | 'subtle' | 'utility';
}

const LoadingSpinner = () => (
  <LoadingSpinnerWrapper alignItems="center">
    <ProgressCircle size="xxSmall" />
  </LoadingSpinnerWrapper>
);

const RawButton: React.FC<ButtonProps & PrivateProps> = memo(
  ({ forwardedRef, isLoading, disabled, ...props }) => {
    return (
      <StyledButton
        className="bd-button"
        {...props}
        disabled={isLoading || disabled}
        ref={forwardedRef}
      >
        {isLoading ? <LoadingSpinner /> : null}
        <ContentWrapper isLoading={isLoading}>
          {!props.iconOnly && props.iconLeft}
          {props.iconOnly}
          {!props.iconOnly && props.children}
          {!props.iconOnly && props.iconRight}
        </ContentWrapper>
      </StyledButton>
    );
  },
);

export const StyleableButton = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <RawButton {...props} forwardedRef={ref} />
));

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, style, ...props }, ref) => <RawButton {...props} forwardedRef={ref} />,
);

const defaultProps = {
  actionType: 'normal' as const,
  isLoading: false,
  mobileWidth: '100%' as const,
  variant: 'primary' as const,
};

Button.displayName = 'Button';
Button.defaultProps = { ...defaultProps };

StyleableButton.displayName = 'StyleableButton';
StyleableButton.defaultProps = { ...defaultProps };
