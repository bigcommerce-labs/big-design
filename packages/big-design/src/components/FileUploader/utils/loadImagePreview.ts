export const loadImagePreview = (file: File): Promise<string | null> => {
  if (!file.type.startsWith('image/')) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      const imageUrl = event.target?.result as string;

      resolve(imageUrl);
    };

    reader.onerror = (error) => {
      // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
      reject(error);
    };

    reader.readAsDataURL(file);
  });
};

export const getImagesPreview = async (files: File[]) =>
  files.reduce(async (prevPromise, file, idx) => {
    const prevResult = await prevPromise;
    const imageUrl = await loadImagePreview(file);

    return { ...prevResult, [idx]: imageUrl };
  }, Promise.resolve<Record<number, string | null>>({}));
