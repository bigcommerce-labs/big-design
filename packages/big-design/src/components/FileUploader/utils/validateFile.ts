import { ValidatorConfig } from '../types';

export const validateFileFormat = (file: File | DataTransferItem, fileFormats = '') => {
  const allowedFormats = fileFormats.split(',').map((format) => format.trim());
  const fileExtension = file.type.split('/').pop()?.toLowerCase() ?? '';

  return allowedFormats.some((format) => {
    if (format === '*' || format === '.*') {
      return true;
    } else if (format.startsWith('.')) {
      return format === `.${fileExtension}`;
    } else if (
      format.startsWith('audio/') ||
      format.startsWith('video/') ||
      format.startsWith('image/')
    ) {
      const mediaTypePrefix = format.split('/')[0];

      return file.type.startsWith(`${mediaTypePrefix}/`);
    }

    return true;
  });
};

export const validateFiles = (files: File[], validators: ValidatorConfig[]) =>
  files.flatMap((file, idx) =>
    validators
      .filter(({ validator }) => !validator(file))
      .map(({ message, type }) => ({ file, fileIdx: idx, message, type })),
  );
