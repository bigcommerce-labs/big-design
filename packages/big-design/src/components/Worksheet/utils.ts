import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
