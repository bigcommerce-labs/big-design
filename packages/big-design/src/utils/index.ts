export * from './typedMemo';
export * from './uniqueId';
export * from './warning';
