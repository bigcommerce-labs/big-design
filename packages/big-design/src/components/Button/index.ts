import { ButtonProps as _ButtonProps } from './Button';

export { Button } from './Button';
export type ButtonProps<T extends 'button' | 'a'> = _ButtonProps<T>;
