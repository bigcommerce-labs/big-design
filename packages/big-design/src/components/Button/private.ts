import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { ButtonProps, StyleableButton as _StyleableButton } from './Button';

export const StyleableButton = _StyleableButton;

export const MessagingButton = styled(_StyleableButton)<ButtonProps>`
  background-color: transparent;
  border: ${({ theme }) => theme.border.none};
  color: ${({ theme }) => theme.colors.secondary70};
  height: auto;
  padding: ${({ theme }) => theme.spacing.none};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary10};
  }

  &[disabled] {
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

MessagingButton.defaultProps = { theme: defaultTheme };
