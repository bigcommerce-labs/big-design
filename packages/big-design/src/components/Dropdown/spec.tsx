import { CheckCircleIcon } from '@bigcommerce/big-design-icons';
import { remCalc } from '@bigcommerce/big-design-theme';
import React, { Fragment } from 'react';
import 'jest-styled-components';

import { fireEvent, render, screen, waitForElement } from '@test/utils';

import { Button } from '../Button';

import { Dropdown } from './';

const onItemClick = jest.fn();

const DropdownMock = (
  <Dropdown
    items={[
      { content: 'Option 1', onItemClick },
      { content: 'Option 2', onItemClick },
      { content: 'Option 3', onItemClick, actionType: 'destructive' },
      { content: 'Option 4', onItemClick, icon: <CheckCircleIcon /> },
    ]}
    toggle={<Button>Button</Button>}
  />
);

const GroupedDropdownMock = (
  <Dropdown
    items={[
      {
        label: 'Label 1',
        items: [
          { content: 'Option 1', onItemClick },
          { content: 'Option 2', onItemClick },
          { content: 'Option 3', onItemClick },
        ],
      },
      {
        label: 'Label 2',
        items: [
          { content: 'Option 4', onItemClick },
          { content: 'Option 5', onItemClick },
          { content: 'Option 6', onItemClick },
        ],
      },
    ]}
    toggle={<Button>Button</Button>}
  />
);

const LineSeparatedGroupedDropdownMock = (
  <Dropdown
    items={[
      {
        label: 'Label 1',
        items: [{ content: 'Option 1', onItemClick }],
      },
      {
        separated: true,
        items: [{ content: 'Option 2', onItemClick }],
      },
    ]}
    toggle={<Button>Button</Button>}
  />
);

test('renders dropdown toggle', () => {
  const { getByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  expect(toggle).toBeInTheDocument();
});

test('dropdown toggle has an id', () => {
  const { getByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  expect(toggle.id).toBeDefined();
});

test('dropdown toggle accepts a custom id', () => {
  const { getByRole } = render(
    <Dropdown items={[{ content: 'Option', onItemClick }]} toggle={<Button id="testId">Button</Button>} />,
  );
  const toggle = getByRole('button');

  expect(toggle.id).toBe('testId');
});

test('dropdown toggle has aria-haspopup', () => {
  const { getByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  expect(toggle.getAttribute('aria-haspopup')).toBe('listbox');
});

test('dropdown toggle has aria-expanded when dropdown menu is open', async () => {
  const { getByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  fireEvent.click(toggle);

  expect(toggle.getAttribute('aria-expanded')).toBe('true');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('renders the dropdown menu closed', () => {
  const { queryByRole } = render(DropdownMock);

  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('opens/closes dropdown menu when toggle is clicked', () => {
  const { getByRole, queryByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  fireEvent.click(toggle);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.click(toggle);
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('dropdown menu has aria-activedescendant', async () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const options = getAllByRole('option');
  expect(getByRole('listbox').getAttribute('aria-activedescendant')).toBe(options[0].id);

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('dropdown menu should have 4 dropdown items', async () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const options = getAllByRole('option');
  expect(options.length).toBe(4);

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('should accept a maxHeight prop', async () => {
  const { getByRole } = render(
    <Dropdown
      items={[
        { content: 'Foo', onItemClick },
        { content: 'Bar', onItemClick },
      ]}
      maxHeight={350}
      toggle={<Button>Button</Button>}
    />,
  );

  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const list = getByRole('listbox');
  expect(list).toHaveStyleRule('max-height', remCalc(350));

  await waitForElement(() => screen.getByRole('option', { name: /foo/i }));
});

test('should default max-height to 250', async () => {
  const { getByRole } = render(DropdownMock);

  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const list = getByRole('listbox');
  expect(list).toHaveStyleRule('max-height', remCalc(250));

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('dropdown items should immediately rerender when prop changes', async () => {
  const { getAllByRole, getByRole, rerender } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  let options = getAllByRole('option');
  expect(options.length).toBe(4);

  rerender(
    <Dropdown
      items={[
        { content: 'Foo', onItemClick },
        { content: 'Bar', onItemClick },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );

  options = getAllByRole('option');
  expect(options.length).toBe(2);

  await waitForElement(() => screen.getByRole('option', { name: /foo/i }));
});

test('first dropdown item should be selected when dropdown is opened', async () => {
  const { getByRole, getAllByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const option = getAllByRole('option')[0];
  expect(option.getAttribute('aria-selected')).toBe('true');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('up/down arrows should change dropdown item selection', async () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const menu = getByRole('listbox');
  const options = getAllByRole('option');

  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  expect(options[1].getAttribute('aria-selected')).toBe('true');

  fireEvent.keyDown(menu, { key: 'ArrowUp' });
  expect(options[0].getAttribute('aria-selected')).toBe('true');

  fireEvent.keyDown(menu, { key: 'ArrowUp' });
  expect(options[3].getAttribute('aria-selected')).toBe('true');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('esc should close menu', () => {
  const { getByRole, queryByRole } = render(DropdownMock);
  const toggle = getByRole('button');

  fireEvent.click(toggle);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.keyDown(getByRole('listbox'), { key: 'Escape' });
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('blurring list should close menu', () => {
  const { getByRole, queryByRole } = render(
    <Fragment>
      {DropdownMock}
      <input type="text" />
    </Fragment>,
  );
  const toggle = getByRole('button');

  fireEvent.click(toggle);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.blur(getByRole('listbox'));
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('home should select first dropdown item', async () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const menu = getByRole('listbox');
  const options = getAllByRole('option');

  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  expect(options[3].getAttribute('aria-selected')).toBe('true');

  fireEvent.keyDown(menu, { key: 'Home' });
  expect(options[0].getAttribute('aria-selected')).toBe('true');
  expect(menu.getAttribute('aria-activedescendant')).toEqual(options[0].id);

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('end should select last dropdown item', async () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const menu = getByRole('listbox');
  const options = getAllByRole('option');

  expect(options[0].getAttribute('aria-selected')).toBe('true');

  fireEvent.keyDown(menu, { key: 'End' });
  expect(options[3].getAttribute('aria-selected')).toBe('true');
  expect(menu.getAttribute('aria-activedescendant')).toEqual(options[3].id);

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('enter should toggle onItemClick', () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const menu = getByRole('listbox');
  const options = getAllByRole('option');

  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  expect(options[1].getAttribute('aria-selected')).toBe('true');

  fireEvent.keyDown(menu, { key: 'Enter' });
  expect(onItemClick).toHaveBeenCalledWith({ content: 'Option 2', onItemClick });
});

test('clicking on dropdown items should toggle onItemClick', () => {
  const { getAllByRole, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const options = getAllByRole('option');
  fireEvent.click(options[1]);
  expect(onItemClick).toHaveBeenCalledWith({ content: 'Option 2', onItemClick });
});

test('dropdown items should be highlighted when moused over', async () => {
  const { getByRole, getAllByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const options = getAllByRole('option');
  fireEvent.mouseOver(options[0]);

  expect(options[0].getAttribute('aria-selected')).toBe('true');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('dropdown menu renders 4 link when passed options of type link', async () => {
  const { container, getByRole } = render(
    <Dropdown
      items={[
        { content: 'Option 1', url: '#', type: 'link' },
        { content: 'Option 2', url: '#', type: 'link' },
        { content: 'Option 3', url: '#', type: 'link' },
        { content: 'Option 4', url: '#', type: 'link' },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );

  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const options = container.querySelectorAll('a');
  expect(options.length).toBe(4);

  options.forEach((option) => {
    expect(option.getAttribute('href')).toBe('#');
  });

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('items renders icons', async () => {
  const { container, getByRole } = render(DropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const svg = container.querySelectorAll('svg');
  expect(svg.length).toBe(1);

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('does not forward styles', async () => {
  const { container, getByRole } = render(
    <Dropdown
      className="test"
      style={{ background: 'red' }}
      items={[{ content: 'Option 1', url: '#', type: 'link' }]}
      toggle={<Button>Button</Button>}
    />,
  );
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(getByRole('listbox')).not.toHaveStyle('background: red');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('renders tooltip with disabled item', async () => {
  const tooltipContent = 'Option with tooltip';
  const tooltipText = 'This is tooltip message';
  const { getByRole, getByText } = render(
    <Dropdown
      items={[
        { content: 'Option 1', onItemClick },
        {
          content: tooltipContent,
          tooltip: tooltipText,
          onItemClick,
          disabled: true,
        },
        { content: 'Option 3', onItemClick, actionType: 'destructive' },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );
  const toggle = getByRole('button');

  fireEvent.click(toggle);
  fireEvent.mouseEnter(getByText(tooltipContent));

  expect(getByText(tooltipText)).toBeInTheDocument();

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test("doesn't render tooltip on enabled item", async () => {
  const tooltipContent = 'Option with tooltip';
  const tooltipText = 'This is tooltip message';
  const { getByRole, getByText, queryByText } = render(
    <Dropdown
      items={[
        { content: 'Option 1', onItemClick },
        {
          content: tooltipContent,
          tooltip: tooltipText,
          onItemClick,
        },
        { content: 'Option 3', onItemClick, actionType: 'destructive' },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );
  const toggle = getByRole('button');

  fireEvent.click(toggle);
  fireEvent.mouseEnter(getByText(tooltipContent));

  expect(queryByText(tooltipText)).not.toBeInTheDocument();

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('no errors expected if all options are disabled', async () => {
  const { getByRole } = render(
    <Dropdown
      items={[
        { content: 'Option 1', onItemClick, disabled: true },
        { content: 'Option 2', onItemClick, disabled: true },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );

  const toggle = getByRole('button');
  expect(() => fireEvent.click(toggle)).not.toThrow();

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('dropdown should have 2 group labels, render uppercased', async () => {
  const { getByRole, getByText } = render(GroupedDropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const label1 = getByText('Label 1');
  const label2 = getByText('Label 2');

  expect(label1).toBeInTheDocument();
  expect(label1).toHaveStyle('text-transform: uppercase');
  expect(label2).toBeInTheDocument();
  expect(label2).toHaveStyle('text-transform: uppercase');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('group labels are grayed out', async () => {
  const { getByRole, getByText } = render(GroupedDropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const label1 = getByText('Label 1');
  const label2 = getByText('Label 2');

  expect(label1).toHaveStyle('color: #8C93AD');
  expect(label2).toHaveStyle('color: #8C93AD');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('group labels are skipped over when using keyboard to navigate options', async () => {
  const { getAllByRole, getByRole } = render(GroupedDropdownMock);
  const toggle = getByRole('button');

  fireEvent.click(toggle);

  const menu = getByRole('listbox');
  const options = getAllByRole('option');

  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  fireEvent.keyDown(menu, { key: 'ArrowDown' });
  fireEvent.keyDown(menu, { key: 'ArrowDown' });

  expect(options[3].getAttribute('aria-selected')).toBe('true');

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('clicking label does not call onItemClick', async () => {
  const { getByRole, getByText } = render(GroupedDropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const label1 = getByText('Label 1');
  fireEvent.mouseOver(label1);
  fireEvent.click(label1);

  expect(onItemClick).not.toHaveBeenCalled();

  await waitForElement(() => screen.getByRole('option', { name: /option 1/i }));
});

test('renders appropriate amount of list items', async () => {
  const { getByRole, container } = render(
    <Dropdown
      items={[
        {
          label: 'Label 1',
          items: [{ content: 'Option 1', onItemClick }],
        },
        {
          items: [{ content: 'Option 2', onItemClick }],
        },
      ]}
      toggle={<Button>Button</Button>}
    />,
  );
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const listItems = await waitForElement(() => container.querySelectorAll('li'));

  expect(listItems.length).toBe(3);
});

test('rendered line separators have correct accessibility properties', async () => {
  const { getByRole, container } = render(LineSeparatedGroupedDropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const hrListItem = await waitForElement(() => container.querySelectorAll('hr')[0].parentElement as HTMLElement);
  expect(hrListItem.getAttribute('aria-hidden')).toBe('true');
  expect(hrListItem.getAttribute('tabindex')).toBe('-1');
});

test('rendered line separators cannot be focused on', async () => {
  const { getByRole, container } = render(LineSeparatedGroupedDropdownMock);
  const toggle = getByRole('button');
  fireEvent.click(toggle);

  const hrListItem = await waitForElement(() => container.querySelectorAll('hr')[0].parentElement as HTMLElement);
  fireEvent.mouseOver(hrListItem);
  expect(document.activeElement).not.toEqual(hrListItem);
});
