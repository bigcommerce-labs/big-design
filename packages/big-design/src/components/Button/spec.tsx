import 'jest-styled-components';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';

import { PlusIcon } from '../Icons';

import { Button } from './index';
import { StyleableButton } from './private';

test('render default button', () => {
  const { container } = render(<Button>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render disabled button', () => {
  const { container } = render(<Button disabled>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render destructive button', () => {
  const { container } = render(<Button actionType="destructive">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render destructive disabled button', () => {
  const { container } = render(
    <Button actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary button', () => {
  const { container } = render(<Button variant="secondary">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary disabled button', () => {
  const { container } = render(
    <Button variant="secondary" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary destructive button', () => {
  const { container } = render(
    <Button variant="secondary" actionType="destructive">
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary destructive disabled button', () => {
  const { container } = render(
    <Button variant="secondary" actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle button', () => {
  const { container } = render(<Button variant="subtle">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle disabled button', () => {
  const { container } = render(
    <Button variant="subtle" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle destructive button', () => {
  const { container } = render(
    <Button variant="subtle" actionType="destructive">
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle destructive disabled button', () => {
  const { container } = render(
    <Button variant="subtle" actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render loading button', () => {
  const { container } = render(<Button isLoading={true}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon only button', () => {
  const { container } = render(<Button iconOnly={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon left button', () => {
  const { container } = render(<Button iconLeft={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon right button', () => {
  const { container } = render(<Button iconRight={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon left and right button', () => {
  const { container } = render(
    <Button iconLeft={<PlusIcon />} iconRight={<PlusIcon />}>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('forwards ref', () => {
  const ref = React.createRef<HTMLButtonElement>();

  const { container } = render(<Button ref={ref} />);
  const button = container.querySelector('button');

  expect(button).toBe(ref.current);
});

test('triggers onClick', () => {
  const onClick = jest.fn();

  const { container } = render(<Button onClick={onClick} />);
  const button = container.firstChild as HTMLElement;

  fireEvent.click(button);

  expect(onClick).toHaveBeenCalled();
});

test('does not forward styles', () => {
  const { container } = render(<Button className="test" style={{ background: 'red' }} />);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('private StyleableButton forwards styles', () => {
  const { container } = render(<StyleableButton className="test" style={{ background: 'red' }} />);

  expect(container.getElementsByClassName('test').length).toBe(1);
  expect(container.firstChild).toHaveStyle('background: red');
});

test('render only icon only with left and right icons button', () => {
  const plusIcon = <PlusIcon data-testid="icon-only" />;
  const { getAllByTestId } = render(
    <Button iconLeft={plusIcon} iconOnly={plusIcon} iconRight={plusIcon}>
      Button
    </Button>,
  );

  expect(getAllByTestId('icon-only').length).toBe(1);
});
