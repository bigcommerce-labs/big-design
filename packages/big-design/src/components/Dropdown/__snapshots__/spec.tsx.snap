// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the dropdown menu open 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  cursor: default;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.25rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-width: 16rem;
  outline: none;
  padding: 0 1rem;
}

.c0[aria-selected='true'] {
  font-weight: 600;
}

.c0[data-highlighted='true'] {
  background-color: #F6F7FC;
}

.c0 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #313440;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  margin: 0 -1rem;
  padding: 0 1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

<li
  class="c0"
  data-highlighted="false"
  id="dropdown_0-item-3"
  role="menuitem"
  tabindex="-1"
  value="3"
>
  Option
</li>
`;
